### PR TITLE
PowerShell task should configure logging streams

### PR DESCRIPTION
**Pull request by bradselw**
**Task name**: PowerShellV2

**Description**: PowerShell [has several output streams](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_output_streams), many of which are used for logging. PowerShell supports different actions when objects are written to each of these streams, such as stopping execution, printing a message, or ignoring it. For example, the `Stop` action tells PowerShell to display a message and stop execution.

While actions can be specified in scripts, a more common pattern is for the caller to specify the preferred "verbosity" of the script by setting the corresponding "preference" variable to a default action, to be used in cases where one is not already provided. For example, the PowerShellV2 task does this with the error stream: It allows the user [to select what action](https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/PowerShellV2/task.json#L78-L90) they want to be taken (i.e. `Stop`, `Continue`, or `SilentlyContinue`) when an object is written to the error stream, by [setting the `$ErrorActionPreference` variable](https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/PowerShellV2/powershell.ps1#L51) at the beginning of the script.

Unfortunately, the PowerShellV2 task currently only gives users control over the error stream. All other streams are handled by PowerShell's default settings, which are rather limited and not suited for automation. Specifically, by default PowerShell suppresses the information and verbose streams, both of which provide crucial logging. The [information stream](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_output_streams#information-stream) is "intended to provide messages that help a user understand what a script is doing", while the [verbose stream](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_output_streams#verbose-stream) is "intended for messages that help users troubleshoot commands as they are run ... from a script." The PowerShellV2 task is meant to run on an agent in an automated environment that the user has limited access to, which makes the logging provided by these streams all the more vital.

This pull request addresses this gap in the PowerShellV2 task's logging in the follow ways:
1. It configures the information stream to log all messages by default
2. It configures the verbose stream to log all messages on an opt-in basis
   - Specifically, the verbose stream will log messages when the pipeline's `System.Debug` variable is set to true. This is an established pattern leveraged across [other tasks](https://github.com/microsoft/azure-pipelines-tasks/blob/master/Tasks/AzurePowerShellV5/AzurePowerShell.ps1#L63-L65).
3. It explicitly defines the action to be taken for the warning stream
   - Hard-coded to always print warnings, instead of relying on PowerShell's defaults

**Documentation changes required:** No.

**Added unit tests:** Yes, updated existing unit tests.

**Attached related issue:** Yes: #15053

**Checklist**:
- [x] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [x] Checked that applied changes work as expected

### CODE REVIEW

**Review by mjroghelia — APPROVED** [review]
This makes sense to me, but I'll leave it to @anatolybolshakov for the final approval.

**Comment by bradselw** [comment]
@EzzhevNikita I pushed a fix for the failing tests. Would you mind kicking off another run?

File: Tasks/PowerShellV2/task.loc.json
Comment (anatolybolshakov) [inline Tasks/PowerShellV2/task.loc.json:20]: Could you please update version to 2.191.0 instread - since current sprint is 191?

File: Tasks/PowerShellV2/task.json
Comment (anatolybolshakov) [inline Tasks/PowerShellV2/task.json:20]: The same here also

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks for contribution!
Could you please take a look at the quick comment above?

**Comment by bradselw** [comment]
@anatolybolshakov I have updated the task version number as requested, thanks for the feedback! Who else needs to sign off before we can merge?

File: Tasks/PowerShellV2/powershell.ps1
Comment (alexander-smolyakov) [inline Tasks/PowerShellV2/powershell.ps1:60]: Nit: According to PowerShell documentation, the default value for `$WarningPreference` variable is [already set](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables?view=powershell-7.1#warningpreference) to `Continue`. Could it make sense to remove it then?

**Review by alexander-smolyakov — APPROVED** [review]
LGTM and please take a look at the comment.

Also tested these changes manually, found no issues.
Link to the related test pipeline: https://v-alsmo.visualstudio.com/Test%20Playground/_build/results?buildId=1316&view=results

File: Tasks/PowerShellV2/powershell.ps1
Comment (anatolybolshakov) [inline Tasks/PowerShellV2/powershell.ps1:59]: Could it maybe make sense also to make these parameters adjustable via task inputs? This could allow to avoid additional noise by default for some cases, also there could be some possible breaking changes - when output for these streams will contain vso commands which will be interpreted and executed by the agent.

File: Tasks/PowerShellV2/powershell.ps1
Comment (bradselw) [reply Tasks/PowerShellV2/powershell.ps1:60]: Yes, the default for `$WarningPreference` is `Continue`. I opted to explicitly set it here (rather than implicitly rely on PowerShell's defaults) for clarity and completeness.

**Comment by anatolybolshakov** [comment]
@bradselw could you please take a look at the comments?

**Comment by bradselw** [comment]
@anatolybolshakov I have made these preferences adjustable via task inputs. A few notes:

- `$VerbosePreference` no longer keys off of `System.Debug`, but uses the task input instead (defaults to `SilentlyContinue`)
- `$InformationPreference` still defaults to `Continue` via the task input default (I still believe this is the correct default behavior)
- I have moved the `$ErrorActionPreference` task input into a group called "Preference Variables" (along with the task inputs for all the other preference variables)
- I have updated the loc strings using the same translations from PowerShell's official documentation and existing patterns

Now that I am adding additional inputs to the task, does this mean we will need to update the task documentation? Also, how does this impact the `powershell` YAML keyword?

File: Tasks/PowerShellV2/Strings/resources.resjson/de-DE/resources.resjson
Comment (anatolybolshakov) [inline Tasks/PowerShellV2/Strings/resources.resjson/de-DE/resources.resjson:22]: Could you please remove updates for localized files and leave English only - since we have separate flow for translation?

File: Tasks/PowerShellV2/powershell.ps1
Comment (anatolybolshakov) [inline Tasks/PowerShellV2/powershell.ps1:47]: Could you please move this repeating pattern to separate cmdlet?

File: Tasks/PowerShellV2/powershell.ts
Comment (anatolybolshakov) [inline Tasks/PowerShellV2/powershell.ts:32]: The same here also, but for function 😃

File: Tasks/PowerShellV2/task.loc.json
Comment (anatolybolshakov) [inline Tasks/PowerShellV2/task.loc.json:20]: Nit: could you please bump it to 192 as current sprint?

**Comment by anatolybolshakov** [comment]
@bradselw thanks! Yes, it make sense to update docs also - you can open the PR in the [repo](https://github.com/MicrosoftDocs/azure-devops-docs).
This should be applicable for powershell alias also - since last task version is used for this.
Could it make sense maybe to enable verbose output for debug mode also?

File: Tasks/PowerShellV2/task.json
Comment (alexander-smolyakov) [inline Tasks/PowerShellV2/task.json:31]: @bradselw @anatolybolshakov

At the moment UI of the task looks like this: 
image

I would like to suggest another way of how we can introduce this feature:
1) By default `Preference Variables`  group will be not visible
2) Add "Override preference variables" toggle to `Advanced` group 
3) By default"Override preference variables" toggle will be set to `false`
4) If the customer toggle `Override preference variables` the `Preference Variables` group will be displayed

So the final flow will look like this:
UI suggestion

Also, if we will keep related logic behind the `Override preference variables` toggle it will allow us to avoid unnecessary injection of `preference variables `with the current default values and save the default behavior of the task.

**Comment by alexander-smolyakov** [comment]
@bradselw it seems there are some problems with the unit test would mind taking a look?

File: Tasks/PowerShellV2/task.json
Comment (anatolybolshakov) [reply Tasks/PowerShellV2/task.json:31]: Could it make sense to add new 'Default' value for these new inputs - to be able to set it as current default value considered by Powershell, and set it up by default? 
@alexander-smolyakov let's quickly discuss it with team also

File: Tasks/PowerShellV2/Strings/resources.resjson/fr-FR/resources.resjson
Comment (bradselw) [inline Tasks/PowerShellV2/Strings/resources.resjson/fr-FR/resources.resjson:9]: Not sure why this is showing up as different.

File: Tasks/PowerShellV2/task.json
Comment (bradselw) [reply Tasks/PowerShellV2/task.json:31]: I'll just point out that the original authors of this task saw fit to explicitly override PowerShell's default action preference for the `Error` stream (which is `Continue`) to `Stop`, because the script is running in an automated environment and you want errors in the script to fail the task. This PR does the same thing for the `Information` stream: because the script is running in an automated environment, any information should be printed to the console instead of ignored.

Comment (bradselw) [reply Tasks/PowerShellV2/task.json:31]: If we were to toggle preferences based on `Override preference variables`, we would break pipelines that rely on the task's default for `ErrorActionPreference`. As a reminder: the task's default for `ErrorActionPreference` is `Stop`, while PowerShell's default is `Continue`.

**Comment by bradselw** [comment]
@alexander-smolyakov and @anatolybolshakov I have addressed most of your feedback. Would you mind kicking off another /azp run?

**Comment by bradselw** [comment]
@anatolybolshakov I have a PR open to update the docs here: https://github.com/MicrosoftDocs/azure-devops-docs/pull/11207

File: Tasks/PowerShellV2/task.json
Comment (anatolybolshakov) [reply Tasks/PowerShellV2/task.json:31]: @bradselw got it, thanks! Could you please add the 'Default' value then for new steams like  `Information` - to keep previous task behavior by default? (default Powershell value)
For `Error` - probably it make sense just leave default value it as is (although Default option would be also useful for consistency)

**Comment by anatolybolshakov** [comment]
@bradselw thanks! Could you please check current comments?

**Review by anatolybolshakov — CHANGES_REQUESTED** [review]
Could you please check comments above?

**Comment by bradselw** [comment]
Sorry about the delay. I had to focus on another project for the past month.

@anatolybolshakov I have made the requested changes. @alexander-smolyakov could we kick off another /azp run?

**Review by anatolybolshakov — APPROVED** [review]
LGTM, thanks!
Tested it locally with different preference values - works fine for me also.

**Comment by anatolybolshakov** [comment]
@bradselw just one final comment - could you please bump task version to 2.194.0 - since for 193 sprint tasks are already shipped to be rolled out?

**Review by alexander-smolyakov — APPROVED** [review]
LGTM, thanks!

**Comment by bradselw** [comment]
@anatolybolshakov done. What are the next steps for getting this rolled out?

**Comment by bradselw** [comment]
PR to update the docs: https://github.com/MicrosoftDocs/azure-devops-docs/pull/11207

@anatolybolshakov, @alexander-smolyakov what are the next steps here?

**Comment by anatolybolshakov** [comment]
@bradselw thanks! We are proceeding with merging at the moment

**Comment by bradselw** [comment]
Thanks @anatolybolshakov. When will this become available broadly? I am trying to decide when to merge the changes to the docs.

**Comment by anatolybolshakov** [comment]
@bradselw it will be rolled out during 195 sprint - per [calendar](https://whatsprintis.it/calendar).

**Pull request merged**
